### PR TITLE
Fix url contains /%2F can't set path correctly

### PR DESCRIPTION
**Pull request by morefreeze**
### Issues associated with this PR

#1651 

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

**Comment by morefreeze** [comment]
@antJack PTAL

File: pkg/stream/http/stream.go [hunk 1]
```diff
@@ -931,8 +931,19 @@ func buildUrlFromCtxVar(ctx context.Context) string {
...
+			// so that /foo/%2Fbar will be /foo/bar(path), but unescapedPath will be /foo//bar
+			// finally it goes here. We need try fasthttp.URI.SetPath and compare them again.
+			// TODO: This if(except else) can be removed if fasthttp fix unescape bug later.
+			u := &fasthttp.URI{}
+			u.SetPath(pathOriginal)
+			fasthttpPath := string(u.Path())
+			if path == fasthttpPath {
+				res = pathOriginal
+			} else {
+				u := url.URL{Path: pathOriginal}
```
Comment (antJack) [inline pkg/stream/http/stream.go:944]: 1. 这里得用 path，不应该是 pathOriginal
2. 最好把 fasthttp 这块代码再突出一下，单独拆个函数或其它方式，u := url.URL{Path:} 不放这么深

File: pkg/stream/http/stream_test.go [hunk 2]
```diff
@@ -51,24 +51,29 @@ func TestBuildUrlFromCtxVar(t *testing.T) {
 			"/home/sa%25%36%64ple",
 		},
 		{
-			"/home/sa%6dple",
+			"/home/sample",
 			"/home/sample%20",
-			"/home/sa%256dple",
+			"/home/sample%2520",
```
Comment (antJack) [inline pkg/stream/http/stream_test.go:56]: 老的测试用例看看是不是可以不用改噢，这其实也是之前别人提问题添加的用例

Comment (morefreeze) [reply pkg/stream/http/stream_test.go]: 这是失误，之前的修复没恢复，当时按pathOriginal那么写的

**Comment by morefreeze** [comment]
@antJack PTAL again

**Review by antJack — DISMISSED** [review]
LGTM

File: pkg/stream/http/stream.go [hunk 3]
```diff
@@ -979,3 +984,9 @@ func (cm *contextManager) next() {
 	// new stream-level context based on connection-level's
 	cm.curr = mbuffer.NewBufferPoolContext(mosnctx.Clone(cm.base))
 }
+
+func fasthttpPath(pathOriginal string) string {
```
Comment (nejisama) [inline pkg/stream/http/stream.go:988]: 这里写一个benchmark看看呢，这个函数看上去是会产生临时对象的。
但是如果直接比较 str == string(byte)  是可能有优化的

Comment (morefreeze) [reply pkg/stream/http/stream.go:988]:
```go
func s2() string {
	s2 := []byte("string to compare B")
	return string(s2)
}

var result interface{}

//BenchmarkStringsCompareDifferent compare 2 strings.
//https://golang.org/pkg/strings/#Compare
func BenchmarkStringsCompareDifferent(b *testing.B) {

	s1 := "string to compare A"

	b.ResetTimer()
	b.ReportAllocs()
	var r bool
	for n := 0; n < b.N; n++ {
		r = s1 == s2()
	}

	result = r
}

func s2byte() []byte {
	s2 := []byte("string to compare B")
	return s2
}

func BenchmarkStringsCompareDifferent2(b *testing.B) {

	s1 := "string to compare A"

	b.ResetTimer()
	b.ReportAllocs()
	var r bool
	for n := 0; n < b.N; n++ {
		r = s1 == string(s2byte())
	}

	result = r
}

```
测了下 str == string(byte) 会快一倍左右，我改下

**Comment by morefreeze** [comment]
@nejisama @antJack PTAL   squash下commit

**Comment by morefreeze** [comment]
@nejisama @antJack PTAL again 相关讨论可以放在原来issue里，这个mr看看是不是可以合

**Review by antJack — APPROVED** [review]
LGTM

**Review by nejisama — APPROVED** [review]

**Pull request merged**